src/Tokenizer/Analyzer/FullyQualifiedNameAnalyzer.php

```diff
@@ -0,0 +1,105 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of PHP CS Fixer.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *     Dariusz Rumiński <dariusz.ruminski@gmail.com>
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace PhpCsFixer\Tokenizer\Analyzer;
+
+use PhpCsFixer\Tokenizer\Analyzer\Analysis\NamespaceAnalysis;
+use PhpCsFixer\Tokenizer\Analyzer\Analysis\NamespaceUseAnalysis;
+use PhpCsFixer\Tokenizer\Tokens;
+
+/**
+ * @internal
+ */
+final class FullyQualifiedNameAnalyzer
+{
+    private Tokens $tokens;
+
+    /**
+     * @var list<NamespaceAnalysis>
+     */
+    private array $namespaceAnalyses = [];
+
+    /**
+     * @var array<int, list<NamespaceUseAnalysis>>
+     */
+    private array $namespaceUseAnalyses = [];
+
+    public function __construct(Tokens $tokens)
+    {
+        $this->tokens = $tokens;
+    }
+
+    /**
+     * @param NamespaceUseAnalysis::TYPE_* $importType
+     */
+    public function getFullyQualifiedName(string $name, int $indexInNamespace, int $importType): string
```

Comment:
see #7724 and #7679 and #7705

All such logic must be implemented by this PR/`FullyQualifiedNameAnalyzer` class otherwise the result will not be always correct.

Then the resolving logic should be used in `FullyQualifiedStrictTypesFixer` and pass all existing tests.

Also, the resolving should be fast and some caching should be supported.

Reply:
This PR is not touching `FullyQualifiedStrictTypesFixer` and I'd like to keep it, and handle it separately, after https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/7705 is merged.

Reply:
Personally I think that caching and optimising can be done is separate PR as currently we don't have it too. Let's don't add features during refactor. When it comes to `FullyQualifiedStrictTypesFixer` I also would like to see it here, but I am open for doing it later too.

@kubawerlos when it comes to 7704/7705 it's been 4 months without any response from the core team, so we don't even know which approach is welcome 🤷. I would happily rebase my PR if it's the chosen way, but I am not going to work on it unless the decision is made.

Reply:
So, since https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/7704 was closed, we can move forward with https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/7705 and this one, and then refactor `FullyQualifiedStrictTypesFixer` to use `FullyQualifiedNameAnalyzer`, right?